components/playlist/renderer/playlist_render_frame_observer.cc

```diff
@@ -60,4 +65,69 @@ void PlaylistRenderFrameObserver::HideMediaSourceAPI() {
       blink::WebString::FromUTF16(kScriptToHideMediaSourceAPI)));
 }
 
+void PlaylistRenderFrameObserver::InstallMediaDetector() {
+  DVLOG(2) << __FUNCTION__;
+
+  static const char16_t kScriptToDetectVideoAndAudio[] =
+      uR"-(
+    (function() {
+      // Firstly, we try to get find all <video> or <audio> tags periodically,
+      // for a a while from the start up. If we find them, then we attach 
+      // MutationObservers to them to detect source URL.
+      // After a given amount of time, we do this in requestIdleCallback().
+      // Note that there's a global object named |pl_worker|. This worker is
+      // created and bound by PlaylistJSHandler.
+
+      let mutationSources = new Set();
+      const findNewMediaAndObserveMutation = () => {
+          return document.querySelectorAll('video, audio').forEach((mediaNode) => {
+              if (mutationSources.has(mediaNode)) return
+
+              mutationSources.add(mediaNode)
+              pl_worker.onMediaUpdated(mediaNode.src)
+
+              new MutationObserver(mutations => {
+                mutations.forEach(mutation => { pl_worker.onMediaUpdated(mutation.target.src); })
+              }).observe(mediaNode, { attributeFilter: ['src'] })
+          });
+      }
+
+      const pollingIntervalId = window.setInterval(findNewMediaAndObserveMutation);
+      window.setTimeout(() => {
+          window.clearInterval(pollingIntervalId)
+          window.requestIdleCallback(findNewMediaAndObserveMutation)
+          // TODO(sko) We might want to check if idle callback is waiting too long.
+          // In that case, we should get back to the polling style. And also, this
+          // time could be too long for production.
+      }, 20000, pollingIntervalId)
```

Comment:
I'm confused about having both `setInterval` here and `setTimeout`:
1. `setInterval` with no `delay` parameter will cause it to call `findNewMEdiaAndObserverMutation` in a loop with no delay between execution (`delay` defaults to `0`).
2. Then `setTimeout` is called with a delay of 20 seconds. In 20 seconds, the loop started in 1. will be cancelled, and the `findNewMediaAndObserveMutation` function will be scheduled to run 1 more time when the thread is next "idle".

Is this because we want to try to catch the src changing and we want the first version? Why not have a delay on the `setInterval` and just check periodically, e.g. every 1 second? I'm worried about performance here. 

Another alternative you could do is have a MutationObserver on the main document, and get notified of new video elements, and/or monkey-patch `document.createElement` and handle when the element is `video` or `audio`? I think this may be what iOS does?

Reply:
I read this as "Let's poll for 20 seconds and see if we get a video". If we didn't get anything enqueue an idle callback, which will hopefully fire when the document finishes changing things around.

:+1: to the 1 second interval

I don't think patching `document.createElement` will work for us, because we'll miss elements created via `innerHTML` and `document.write` (and whatever other crazy ways you can create elements) but we could definitely use a mutation observer on the body:

This should behave identically, but only do the find in ticks where the DOM changed - dunno if it's better than just checking every second though :laughing: 
```js
const mutationObserver = new MutationObserver(e => {
    // todo: be cleverer here and only do this if video/audio elements were added/removed/src changed
    findNewMediaAndObserveMutation()
})
mutationObserver.observer(document.body, { attributeFilter: ['src'], childList: true, subtree: true })

setTimeout(() => {
    mutationObserver.disconnect();
    window.requestIdleCallback(findNewMediaAndObserveMutation);
}, 20000);
```

Reply:
Also, out of curiosity, what does the `pollingIntervalId` do as an argument to `setTimeout`? I've never seen it before

```js
setTimeout(() => ..., 20000, pollingIntervalId)
```

Reply:
> setInterval with no delay parameter will cause it to call findNewMEdiaAndObserverMutation in a loop with no delay between execution (delay defaults to 0).

> setTimeout(() => ..., 20000, pollingIntervalId)

Ah, my bad. I made a mistake during refactoring.

---

> setInterval and just check periodically, e.g. every 1 second? I'm worried about performance here.

>  I understand this: We poll for 20 seconds and see if we get a video. If we didn't get anything enqueue an idle callback, 

Yeah, the overall strategy is from cosmetic_js_handler.cc, which is switching between polling and idle-time detection. So What I try to do is polling for the first seconds, then turn it to idle-time detection. The reason why polling first is that it's more likely for videos to be attached in the beginning. 

Does polling is better than rIF in terms of performance?

---

And one more thing, @diracdeltas and I was talking about the performance problem of `MutationObserver` and I guess we can look into it later as this scrpit will be run only on the background/copied web conetents for now. - https://github.com/brave/brave-browser/issues/29488#issuecomment-1502666818

And there's a native API which acts just like `MutationObserver`, so I'm going to try it, before too late.

Reply:
@fallaciousreasoning , I'd like to ask something about your suggestion

```
const mutationObserver = new MutationObserver(e => {
    // todo: be cleverer here and only do this if video/audio elements were added/removed/src changed
    findNewMediaAndObserveMutation()
})
mutationObserver.observer(document.body, { attributeFilter: ['src'], childList: true, subtree: true })
```

Does this mean that one `MutationObserver` will observe the entire DOM tree from the document.body as a root? I'm concerned if it'll cause performance issue mentioned [here](https://github.com/brave/brave-browser/issues/29488#issuecomment-1502666818). And that was the initial idea why I made a multiple mutation observer that observes only "src" attribute. And IIRC, `document.body` wasn't valid at the time this code was executed 😥

Reply:
Yeah, if we're concerned about perf the polling approach might be better :smile: Optionally, we could still share the mutation observer if we wanted (and just call `.observe` on all the video/audio elements).